Add support for Metrowerks C/C++ compilers

Nomura-RH opened the pull request:
Hello meson devs!

I added preliminary support for the Metrowerks C/C++ compiler(s). Metrowerks is a company with a focus on providing toolchains targeting embedded systems. It was acquired in 2005 by Freescale Semiconductors, which itself was acquired by NXP Semiconductors in 2015. However, the tools are still named after Metrowerks (they are prefixed with 'mw'), so I believe Meson should do the same.

While all Metrowerks compilers share a common CLI, I decided to focus on the Embedded ARM and PowerPC toolchains for the time being. They contain a C/C++ compiler, a standalone assembler, and a linker that acts both as a static and dynamic linker. The linker is compatible with GNU AR archives, but does not support thin archives.

I suppose this isn't perfect on first try (plus I don't usually develop in Python), so just let me know what to adapt :)

Best regards, Nomura

bruchar1 commented on mesonbuild/compilers/detect.py:
```diff
@@ -572,6 +577,36 @@ def sanitize(p: str) -> str:
+                cls = c.MetrowerksCCompilerEmbeddedPowerPC if lang == 'c' else cpp.MetrowerksCPPCompilerEmbeddedPowerPC
+                lnk = linkers.MetrowerksLinkerEmbeddedPowerPC
+
+            mwcc_ver_match = re.search(r'Version (\d+)\.(\d+)\.?(\d+)? build (\d+)', out)
```
```suggestion
            mwcc_ver_match = re.search(r'Version (\d+)\.(\d+)(?:\.(\d+))? build (\d+)', out)
```
(?: ... ) is a non-capturing group. Because I guess you dont want the number to finish with a `.`

Nomura-RH replied:
If the version doesn't have a third number, the final dot also won't be in the compiler's output, so I think the regex is ok as is?

bruchar1 replied:
Yes. It is just that it would also match `1.2.`, but probably it cannot happen....

bruchar1 commented on mesonbuild/compilers/detect.py:
```diff
@@ -572,6 +577,36 @@ def sanitize(p: str) -> str:
+            if ld is not None:
+                _, o_ld, _ = Popen_safe(ld + ['--version'])
+
+                mwld_ver_match = re.search(r'Version (\d+)\.(\d+)\.?(\d+)? build (\d+)', o_ld)
```
```suggestion
                mwld_ver_match = re.search(r'Version (\d+)\.(\d+)(?:\.(\d+))? build (\d+)', o_ld)
```

bruchar1 commented on mesonbuild/compilers/detect.py:
```diff
@@ -572,6 +577,36 @@ def sanitize(p: str) -> str:
+
+                mwld_ver_match = re.search(r'Version (\d+)\.(\d+)\.?(\d+)? build (\d+)', o_ld)
+                assert mwld_ver_match is not None, 'for mypy'  # because mypy *should* be complaning that this could be None
+                linker_version = '.'.join([x for x in mwld_ver_match.groups() if x is not None])
```
```suggestion
                linker_version = '.'.join(x for x in mwld_ver_match.groups() if x is not None)
```

You dont need an intermediate list. Just a generative expression. `filter()` would work as well, but I'm not sure everybody like it...

bruchar1 commented on mesonbuild/compilers/mixins/metrowerks.py:
```diff
@@ -0,0 +1,224 @@
+
+    def compute_parameters_with_absolute_paths(self, parameter_list: T.List[str], build_dir: str) -> T.List[str]:
+        for idx, i in enumerate(parameter_list):
+            if i[:9] == '-I':
```
There must be something wrong... a nine chars string cannot be equal to a two chars string!

Nomura-RH replied:
Whoops, that's a leftover from another compiler class I used as a base. Not sure what it's used for.

eli-schwartz replied:
It matches on --girdir= (vala) or /LIBPATH: (MSVC) or -include=

The purpose of this function is so that the IDE introspection `*.json` files, and specifically intro-targets.json, have a `"parameters": []` entry which specifies each member of parameter_list (the actual compiler arguments run to compile a given target) with `-I<relative-path>` converted to `-I<absolute-path>`. Mismatching here is "harmless" and results in a no-op and technically, reduced functionality.

It looks like you probably copied this from the compcert compiler, which has been broken in this fashion ever since its original implementation in #7674 -- I would accept a second patch that corrects this in the compcert mixin. :)


(Please, either way, update your metrowerks code here, to use `i[:2]`.)

eli-schwartz replied:
(The compcert compiler presumably copied it from some code that used MSVC style or something, updated the string, but didn't update the character count.)

bruchar1 commented:
I know nothing about Metrowerks, but I commented about Python syntax.

Nomura-RH commented:
Since this toolchain may not be very well-known, here's a manual for the ARM-targeting version that is publicly available on NXP's website: https://www.nxp.com/docs/en/reference-manual/CWMCUKINCMPREF.pdf

eli-schwartz commented on mesonbuild/compilers/mixins/metrowerks.py:
```diff
@@ -0,0 +1,224 @@
+
+    # These compilers can actually invoke the linker, but they choke on
+    # linker-specific flags. So it's best to invoke the linker directly
+    INVOKES_LINKER = False
```
Interesting, there's no options similar to `-Wl,` here?

eli-schwartz commented on mesonbuild/compilers/detect.py:
```diff
@@ -572,6 +577,36 @@ def sanitize(p: str) -> str:
+
+                linker = lnk(ld, for_machine, version=linker_version)
+            else:
+                raise EnvironmentException(f'Failed to detect linker for \'{cls.id}\' compiler. Please update your cross file(s).')
```
Python can internally do the quoting here.

```suggestion
                raise EnvironmentException(f'Failed to detect linker for {cls.id!r} compiler. Please update your cross file(s).')
```

eli-schwartz commented on mesonbuild/compilers/detect.py:
```diff
@@ -572,6 +577,36 @@ def sanitize(p: str) -> str:
+                cls = c.MetrowerksCCompilerEmbeddedPowerPC if lang == 'c' else cpp.MetrowerksCPPCompilerEmbeddedPowerPC
+                lnk = linkers.MetrowerksLinkerEmbeddedPowerPC
+
+            mwcc_ver_match = re.search(r'Version (\d+)\.(\d+)\.?(\d+)? build (\d+)', out)
```
How does the usual search_version function handle this compiler's version strings?

Nomura-RH replied:
It detects the major and minor version, but not the build number. This is not surprising, because the version strings are in the format: Version X.Y build ZZZZ and not X.Y.Z

eli-schwartz reviewed (commented):
Thanks for working on this! I'll take a closer look soon, but I noticed a couple things on the first pass...

Incidentally, can you confirm that the test suite more or less seems to be doing the right thing, when you execute `./run_cross_test.py --cross-only` and with a suitable cross definition in cross/?

Nomura-RH commented:
> Thanks for working on this! I'll take a closer look soon, but I noticed a couple things on the first pass...
> 
> Incidentally, can you confirm that the test suite more or less seems to be doing the right thing, when you execute `./run_cross_test.py --cross-only` and with a suitable cross definition in cross/?

@eli-schwartz Unfortunately the linker requires a linker script to link an application, so it gets stuck on the very first test...

eli-schwartz commented:
Is this something generic enough that you can create one which is suitable for plugging into the cross file as the link args? Do you need a linker script fine-tuned to the application you're building?

Nomura-RH commented:
> Is this something generic enough that you can create one which is suitable for plugging into the cross file as the link args? Do you need a linker script fine-tuned to the application you're building?

Yes, I was able to make a generic linker script that works with the tests. I am attaching the results.
[test_log.txt](https://github.com/mesonbuild/meson/files/11093576/test_log.txt)

eli-schwartz commented:
Neat! Looking at the results, there is some low-hanging fruit:

```
During: install
Reason: 
Extra file usr\bin\prog.nef found.
Expected file usr\bin\prog missing.
```

This is a problem for Windows too (.exe file extensions) so you can fine-tune that here:
https://github.com/mesonbuild/meson/blob/728c8dce245a4df095e5acab42db769d79efd1ba/run_project_tests.py#L214-L216

Overall results:
- Total passed tests:  181
- Total failed tests:  63
- Total skipped tests: 18

That's very promising and indicates this is pretty useful to start with. You may want to also investigate the exact cause of each of those failures, and try to fix whatever issues are causing it -- but that's up to you, the perfect is the enemy of the good...

Fair warning: we are in feature freeze for 1.1.0-rc1, so this PR will probably not be merged until after the final release is cut. It's unlikely that this will need rebasing between now and then, for the same reason.

Nomura-RH commented:
> 
> Overall results:
> 
> * Total passed tests:  181
> * Total failed tests:  63
> * Total skipped tests: 18
> 
> That's very promising and indicates this is pretty useful to start with. You may want to also investigate the exact cause of each of those failures, and try to fix whatever issues are causing it -- but that's up to you, the perfect is the enemy of the good...
> 
> Fair warning: we are in feature freeze for 1.1.0-rc1, so this PR will probably not be merged until after the final release is cut. It's unlikely that this will need rebasing between now and then, for the same reason.

Looking into the reasons for failure, it seems that most of the time the compiler is just not finding some standard library headers. 

Another not-so-great thing to know about this toolchain is that, while it comes together with its own implementations of the C and CPP standard libs, it cannot find their headers on its own. The user has to either set certain environment variable to point at their location, or pass include paths to the compiler. 

Point being, a lot of failures are probably due to my machine's setup, not the code. Have to check though.

> Fair warning: we are in feature freeze for 1.1.0-rc1, so this PR will probably not be merged until after the final release is cut. It's unlikely that this will need rebasing between now and then, for the same reason.

Do you have a rough estimate of when the final release will be ready? I am considering porting an open-source project I am maintaining from Make to Meson, and I don't think it's a good idea to have users install my fork for the purpose of building that specific project alone.

eli-schwartz commented:
We will probably have an rc2 next week, and hopefully the final release will be the week after.

eli-schwartz reviewed (commented):
Meson 1.1.0 is now released and the master branch is unfrozen for development. Looking forward to landing this soon!

eli-schwartz commented:
@Nomura-RH by the way, if you feel like it I would be totally okay with you checking in your cross file to the `cross/` directory. It wouldn't be directly useful in our CI, but anyone else who happens to be playing with the metrowerks compiler could use it to get started with running the testsuite... possibly even you in future contributions. :)

Feel free to ignore this comment, though.

eli-schwartz commented on run_project_tests.py:
```diff
@@ -212,7 +212,10 @@ def get_path(self, compiler: str, env: environment.Environment) -> T.Optional[Pa
             return p.with_suffix(suffix)
         elif self.typ == 'exe':
-            if env.machines.host.is_windows() or env.machines.host.is_cygwin():
+            if any('mwcc' in s for s in env.lookup_binary_entry(MachineChoice.HOST, 'c')) or \
+               any('mwcc' in s for s in env.lookup_binary_entry(MachineChoice.HOST, 'cpp')):
```
Can we use canonical_compiler for this check? Should be a lot simpler.

Nomura-RH commented on mesonbuild/linkers/linkers.py:
```diff
@@ -1554,3 +1576,46 @@ def get_allow_undefined_args(self) -> T.List[str]:
+        return []
+
+    def get_output_args(self, target: str) -> T.List[str]:
+        return [f'-o {target}']
```
Mistake on my end, should've split the args into two strings

Nomura-RH commented:
> @Nomura-RH by the way, if you feel like it I would be totally okay with you checking in your cross file to the `cross/` directory. It wouldn't be directly useful in our CI, but anyone else who happens to be playing with the metrowerks compiler could use it to get started with running the testsuite... possibly even you in future contributions. :)
> 
> Feel free to ignore this comment, though.

To make the cross file useable I would have to check in the linker script as well. Is that ok?

eli-schwartz commented:
Sure, just stick it in cross/ too. It won't interfere with general project maintenance either way, and it's just as useful for future contributors to metrowerks support as the cross file itself. :)

Nomura-RH commented:
Need help with a few more issues before I submit the cross file:

1. link_args in the cross file are being ignored during static linking. This happens even when I define the args on a per-target basis.

2. Libraries are getting enclosed within single quotes on the link command line. This breaks the build. Currently I rely on a postconf script to remove the quotes from the build.ninja file

eli-schwartz commented:
1. The implication here is that when bundling up an "ar" archive of object files, the user needs the ability to control the arguments used? This isn't currently factored into how meson works, because it's not necessary for either the average unix static linker or for Microsoft's link.exe -- and certainly, trying to pass options valid for binutils ld / lld to binutils `ar` won't ever work. Is this truly necessary for Metrowerks? If so, it may need special casing. For example in ninjabackend.py, the function generate_link() will add `self.build.get_project_link_args()` (for the similarly named meson.build function) and `linker.get_target_link_args()` which provides the `link_args` kwarg of a target... but only when the result is a dynamic library. Should that be "is a dynamic library or the linker is the Metrowerks one"?

2. Oh hmm, I remember now that you asked about this but I forgot to reply. I wonder why this is needed. For macOS/Linux the single quotes are interpreted by the unix shell and the program just sees an array of `argv` arguments, but for Windows it is the job of every single program to take the command line as a single string and do its own breaking into argv "words". Including the processing of quotes. Does the Windows build of Metrowerks have a bug here?

Nomura-RH commented:
> 1. The implication here is that when bundling up an "ar" archive of object files, the user needs the ability to control the arguments used? This isn't currently factored into how meson works, because it's not necessary for either the average unix static linker or for Microsoft's link.exe -- and certainly, trying to pass options valid for binutils ld / lld to binutils `ar` won't ever work. Is this truly necessary for Metrowerks? If so, it may need special casing. For example in ninjabackend.py, the function generate_link() will add `self.build.get_project_link_args()` (for the similarly named meson.build function) and `linker.get_target_link_args()` which provides the `link_args` kwarg of a target... but only when the result is a dynamic library. Should that be "is a dynamic library or the linker is the Metrowerks one"?

It's not really necessary, I'm just trying to turn off some useless warnings. I couldn't get --warnlevel=0 to play nice with the static linker in my PR, so I thought I could achieve what I want by passing the needed args to static library targets. EIther way, if the user is not supposed to control these args, you might want to update the documentation. It mentions link_args as a valid keyword argument for static_library(): https://mesonbuild.com/Reference-manual_functions.html#static_library

> 2. Oh hmm, I remember now that you asked about this but I forgot to reply. I wonder why this is needed. For macOS/Linux the single quotes are interpreted by the unix shell and the program just sees an array of `argv` arguments, but for Windows it is the job of every single program to take the command line as a single string and do its own breaking into argv "words". Including the processing of quotes. Does the Windows build of Metrowerks have a bug here?

Metrowerks is a ...quirky toolchain to say the least. It can break down the args correctly if they are enclosed within double quotes, but not single quotes. I don't even think this is a bug, the developers probably didn't bother.

eli-schwartz commented:
> EIther way, if the user is not supposed to control these args, you might want to update the documentation. It mentions link_args as a valid keyword argument for static_library(): https://mesonbuild.com/Reference-manual_functions.html#static_library

This does seem like a bug, and we should fix things to warn about it even.

> Metrowerks is a ...quirky toolchain to say the least. It can break down the args correctly if they are enclosed within double quotes, but not single quotes. I don't even think this is a bug, the developers probably didn't bother.

Hmm, double checking I think that meson is *supposed* to double quote arguments on Windows, and single-quote them on unix. e.g. in ninjabackend.py the `cmd_quote()` function for Windows, and in mesonbuild/utils/universal.py, the `quote_arg()` function. I wonder why single quotes are ending up here.

Nomura-RH commented:
Alright, I'll put aside figuring out these issue for the time being. Besides the cross file, is there anything else you need me to fix in order to get this merged?

eli-schwartz commented:
https://mesonbuild.com/Machine-files.html#constants

You can do "layering", so the cross file needs to be used at the same time as another cross file which defines a base variable for the directory.

I seem to recall there was a proposal somewhere to add a way to refer to the directory with the cross file, but I don't think it ever landed... Hmm.

...

Found it. #9223

Nomura-RH commented:
Alright, the cross file and linker script are in. I also added a release snippet and updated the reference tables with the metrowerks compiler and linker ids.

eli-schwartz commented on docs/markdown/snippets/add_metrowerks_compiler.md:
```diff
@@ -0,0 +1,5 @@
+
+Added support for the Metrowerks for Embedded ARM and Metrowerks for Embedded PowerPC toolchains (https://www.nxp.com/docs/en/reference-manual/CWMCUKINCMPREF.pdf).
+
+The implementation is somewhat experimental. It has been tested on a few projects and works fairly well, but may have issues
```
Sentence is missing a period at the end.

Nomura-RH replied:
Whoopsie. Fixed.

eli-schwartz reviewed (approved):
Thanks! I have exactly one grammar nit in the release note entry but otherwise this should be ready to merge.

eli-schwartz commented:
Scratch that, there are some linter errors too... Should be pretty easy to solve though. :)

Nomura-RH commented:
> Scratch that, there are some linter errors too... Should be pretty easy to solve though. :)

I (hopefully) fixed all of them

eli-schwartz commented on mesonbuild/compilers/cpp.py:
```diff
@@ -889,3 +891,60 @@ def get_option_link_args(self, options: 'KeyedOptionDictType') -> T.List[str]:
+                 linker: T.Optional['DynamicLinker'] = None,
+                 full_version: T.Optional[str] = None):
+        CPPCompiler.__init__(self, ccache, exelist, version, for_machine, is_cross,
+                           info, exe_wrapper, linker=linker, full_version=full_version)
```
```suggestion
                             info, exe_wrapper, linker=linker, full_version=full_version)
```

Lint problem.

eli-schwartz commented on mesonbuild/compilers/detect.py:
```diff
@@ -573,6 +578,37 @@ def sanitize(p: str) -> str:
                 exe_wrap, full_version=full_version, linker=linker)
 
+        if 'Metrowerks C/C++' in out or 'Freescale C/C++' in out:
+            lnk: T.Union[T.Type[linkers.MetrowerksLinkerARM], T.Type[linkers.MetrowerksLinkerEmbeddedPowerPC]]
```
This variable has a different type up above, and due to Reasons, mypy considers it an error to change the type in a different branch. There are two general solutions to this:
- name it lnk2 or something
- raise its definition to be right underneath `cls: T.Union[T.Type[CCompiler], T.Type[CPPCompiler]]`; in this case however, you define it to match any StaticLinker or DynamicLinker, so it covers both Metrowerks and TI.

eli-schwartz reviewed (commented):
We're almost there.

Nomura-RH commented:
> We're almost there.

Fingers crossed 🤞
